### ISSUE TITLE
Drupal Security Updates: Define & Implement Process

### ISSUE THREAD

**Issue by mikemilano** [thinktandem/tandem#68]
Protecting client data might be the highest responsibility we have to our clients. Client data is most easily breached via unpatched security vulnerabilities.

It is not acceptable for security vulnerabilities to exist for sites in our care and we should do our best to reduce the chance of missing a timeline update.

## Optimal Scenario

In a perfect world, anyone in management or engineering would be able to tell, at a glance, that any Drupal instance in our care has avaialble security updates.

## Current Scenario

We currently rely on one developer to manually check and keep Drupal core & contrib secure for sites in our custody. To be clear, this person has to login or run commands against each site to be absolutely sure there are no updates available.

Even with a high caliber developer in charge, this is error prone. It is also not scalable.

## Issues with the Current Method

Security updates could get missed if:
- Checking for updates is forgotten
- Checking for updates isn't done because there is no process or expectation defined
- A site gets missed in the process of checking for updates
- The one responsible for this is out or on vacation
- Drupal misses a module status when retrieving the list of updates
- An update was deployed wrong (to the wrong branch, or an error was not noticed)

## Phases

![drupal-security-updates](https://user-images.githubusercontent.com/1287855/61387372-db414b80-a86a-11e9-92da-eb16f1f33daa.jpg)

We can approach improving process in phases by first by defining a policy and making sure all sites are checked and updated at a regular frequency. Then in later phases we can work towards automation.

Heck, we might even be able to make it into a service at one point, but let's start simple.

### Phase 1: Define policy

First of all we need a policy defined. I believe @pirog may be working on a separate effort to define this based on slack conversation.

> What is your security plan?

Basically, be prepared to answer this as it relates to Drupal Updates when asked this by the CTO of a client.

For example:
- We monitor Drupal Security team for announcements
- Sites are checked for security updates daily
- Security updates are scheduled to be applied once detected

## Phase 2: Make it easier to detect updates

This can actually be done in conjunction with Phase 1.

While it would require someone to maintain drush aliases to all projects in our custody, all sites could be checked with a single command using something like [Drupdates](https://github.com/thinkbean/drupdates)

This would make it less of a hassle to thoroughly check all sites, making it more likely they would get checked on a regular basis.

## Phase 3: Automated checks, manual updates

As a next step, available updates would be checked automatically and perhaps reported to slack each morning.

This would require an operations server where a combination of Platform.sh API, Drush, and Slack API is used to acommodate the requirements.

I estimate about a day or 2 of work to come up with a MVP using something like a Symfony Console command which is executed via cron.

## Phase 4: Improve, automate, etc...

It's a bit premature to plan exactly what we will do here but next steps might include:
- Automating the process of updating modules and creating a pull request
  - There will always be exceptions for when this may not be a good idea
- Make this into a service we can offer developers/agencies

## Ticket Checklist

Make sure you complete the following checklist

- [x] I'm aware of [what a discussion actually entails](https://docs.thinktandem.io/guides/improve-tandem.html#suggesting-a-change)
- [x] I've assigned this task to myself and one of the Keepers of the Faith
- [x] I've **not** set a story point estimation
- [x] I've selected _at most_ two labels that _best_ describe what part of the business this ticket improves eg `sales`, `dev`, etc

**Comment by mikemilano** [comment]
@labboy0276, I opened this discussion based on the Slack conversation a few days ago. This affects you most so it would be great to have your input.

**Comment by pirog** [comment]
@mikemilano this is great. awesome work clearly describing the problem, current state of things and a roadmap that aligns with our goals as an org.

So, as you mentioned above: i'm going to spin up a task (or two) for us to define/clarify

1. Our security policy, eg the high level approach Tandem takes that shouldn't change much over time
2. Our current process for doing security updates eg the current "implementation details" of our policy

I dont think this discussion is necessarily blocked by those two tasks but it might be wise to try and tackle them first to provide better context here. Once i spin them up i'll drop some comments in here and we can discuss how we want to organize our efforts. 

What do you think?

**Comment by labboy0276** [comment]
So @mikemilano I concur we need to do something.  We were a test case for this: https://www.drupal.org/project/evercurrent and it wasn't ready for primetime.  It could be different now and it wasn't a horrible solution.  

Your drupudates solution doesn't seem to work with our current workflow.  When I look at the codes here: https://github.com/thinkbean/drupdates/blob/master/src/Drupdates/Util/DrupalUtil.php it seems you are checking for updates with ```drush pm-updatestatus``` which is outdated in a composer workflow:

```bash
john@tandem:~/lando/twinstar (36/more-fixzz)$ lando drush pm-updatestatus --format=json

In Application.php line 268:
                                                                                                                                                                       
  The pm-updatestatus command was deprecated. Please see `composer show` and `composer update --dry-run`. For security release notification, see `drush pm:security`.  
                                                                                                                                                                      
```

The ```drush pm:security``` && ```composer update --dry-run``` options are kind of lackluster too.  My suggestion is we look back into https://www.drupal.org/project/evercurrent and see how they have come along.

**Comment by pirog** [comment]
@labboy0276 awesome feedback here but FWIW we might not want to dive super deep into implementation details right now. 

A good next step might be sketching out HOW we want this whole thing to work eg "we want to get automatic pull requests in github when updates happen and slack notifications for those PRs" and its scope eg "we want to get security updates for drupal, node, etc" vs WHAT THING  eg "we will use drupdates" will make it happen. Obviously it makes sense to figure out a HOW that fits easily into our current workflow.

**Comment by labboy0276** [comment]
@mikemilano @pirog OK.  The ideal workflow is we get a notification there are updates either by email or on a slack channel, slack channel is preferred.  .  It would be nice if updates were submitted via PRs to a GitHub repo as well if they have a github repo.  There are some edge case clients that fall outside of our workflow and just the email / slack notification is good enough to trigger us to do the updates.

**Comment by pirog** [comment]
@labboy0276 yup something like that sounds good as the end destination. 

Let me work on Phase I of this eg the policy part. Maybe you and @mikemilano can sketch out our current process (@serundeputy might be good to add since hes looked at and implemented some of GitHubs automated security bot solutions for our node deps) and then start to sketch out the phase 2 @mikemilano defined. 

How do you guys feel about that task breakup? Give me the 👍 and i'll try to spin up teh relevant tasks and assign them

**Comment by labboy0276** [comment]
Just putting this on here because IMS uses this to create PRs for Drupal updates via composer:
https://github.com/marketplace/dependabot-preview

**Comment by pirog** [comment]
alright, i've spun up the first two issues to translate this discussion into actionable tasks. With the followups in each and appropriate backlinks to this discussion for context i think the ship has sailed and we can close this out in favor of those
https://github.com/thinktandem/tandem/issues/114
https://github.com/thinktandem/tandem/issues/115
